JP-1431: add parameter extract_height to wfss extraction

nden opened the pull request:
Fixes #4896 / [JP-1431](https://jira.stsci.edu/browse/JP-1431)

nden commented:
@hbushouse Is the `extraction_height` parameter supposed to be a full or half height?

hbushouse commented:
JP-1431 requests that the extraction parameter be a half-height, such that param=N results in a total height of 2N+1.

jdavies-st commented on jwst/assign_wcs/util.py:
```diff
@@ -506,7 +506,8 @@ def get_object_info(catalog_name=None):
                       mmag_extract=99.0,
-                      extract_orders=None):
+                      extract_orders=None,
+                      extract_height=None):
```
If the keyword is `extract_height` it would be clearer if it was the full height of the extraction.  And this is already what the current keyword is.

https://github.com/spacetelescope/jwst/blob/ff9ea4017ebdc62309cf97661a5093f72c402722/jwst/extract_2d/extract_2d.py#L37-L40

If it is to be the half height, then `extraction_half_height` would be clearer to users I think.

jdavies-st commented on jwst/assign_wcs/util.py:
```diff
@@ -562,8 +563,17 @@ def create_grism_bbox(input_model,
         filter_name = input_model.meta.instrument.filter
+        if input_model.meta.instrument.pupil.endswith('R'):
+            # _extraction_height_direction is used if extract_height is not None
+            _extract_height_direction = 'y'
+        else:
+            _extract_height_direction = 'x'
```
The DISPAXIS or `meta.wcsinfo.dispersion_direction` keyword should give the dispersion direction without having to rely on filter name parsing.  Or move the filter name parsing to the dispersion direction function

https://github.com/spacetelescope/jwst/blob/ff9ea4017ebdc62309cf97661a5093f72c402722/jwst/lib/dispaxis.py#L6-L8

hbushouse commented:
> Does this also allow all the grism extractions to have their height set to a specific value, not just SOSS? That could be useful for the WFSS modes as well I suspect.

This PR **IS** for WFSS mode, not SOSS.

jdavies-st commented:
Doh!  _brainfart_

Then the `extract_height` keyword should be consistent with what it is for SOSS, i.e. the full height of the extraction box.  Having the same keyword mean 2 different things with two different datasets using the same code is not good.

nden commented:
@jdavies-st SOSS does not go through extract_2d. The parameter is for TSGRISM mode. But it's a good point. @hbushouse Would it be possible to change the TSGRISM parameter to "half height" as well or does the INS team wants is to be a full height. If so we'll have to have two parameters.

hbushouse commented:
I know the NIRCam TSO folks were specific about the placement and height of the extraction box that they wanted for TSGRISM. If the current height happens to be an odd number, then it too could be expressed as a half-height, like WFSS. But I'm guessing that the default values will be different for TSGRISM and WFSS, so it's not possible to combine into a single parameter, unless we create separate extract_2d parameter files and install them in CRDS. During processing, the appropriate one would be selected based on EXP_TYPE.

nden commented:
In this case I'll create a new parameter `extract_half_hight_grism` and say the rest is out of scope for this PR. If there are better ideas, please speak.

hbushouse commented:
That's fine, but I would use the term "wfss" in the parameter name, instead of the generic "grism", because TSGRISM and WFSS both use a grism. This parameter will be specific to WFSS only. Perhaps "wfss_extract_half_height".

nden commented:
I took the opportunity to refactor `create_grism_bbox` by splitting it into two functions (unfortunately still long), one that performs the computations and one that sets the input parameters. I also added another optional parameter, `wavelength_range`, to `create_grim_bbox` which gives more control over setting the extraction parameters.

hbushouse commented on docs/jwst/extract_2d/main.rst:
```diff
@@ -78,6 +78,80 @@ of ``GrismObjects`` outside of these, the ``GrismObject`` itself can be imported
+reference file which can be retrieved from CRDS. A user may supply a different
+wavelength range by passing `None` to ``reference_files``. In this case the spectral
+orders to be extracted and their corresponding wavelength range will be taken
+from the ``wavelength_range`` parameter.
```
How does a user specify both wavelength ranges AND spectral orders via a single parameter? Or is that explained later?

nden replied:
The `wavelength_range` parameter is a dictionary of 
`{spectral_order: (lam_min, lam_max)}`

hbushouse commented on docs/jwst/extract_2d/main.rst:
```diff
@@ -78,6 +78,80 @@ of ``GrismObjects`` outside of these, the ``GrismObject`` itself can be imported
+  {'wavelengthrange': '/crds/references/jwst/niriss/jwst_niriss_wavelengthrange_0002.asdf'}
+
+Create a list of grism objects for a specified spectral order with a limited
+minimum magnitude, and a specified half height of the extraction box in
```
Should note that the user-specified extraction height is only applied to grism objects with a source type of POINT (non-point sources should use the normal bb limits).

hbushouse commented on docs/jwst/extract_2d/main.rst:
```diff
@@ -133,11 +207,13 @@ For NIRCam and NIRISS WFSS, the ``extract_2d`` step has three optional arguments
+For NIRCam and NIRISS WFSS mode, the ``extract_2d`` step has one optional argument:
+
+  ``--wfss_extract_half_height``
+    int. The cross-dispersion half size, in pixels.
```
Again, note that this only applies to sources tagged in the source catalog as point-like (based on the catalog entry "is_star" - boolean).

nden replied:
will make that check and point this pout in the documentation

nden replied:
The catalog I have lists `is_star` as float, not boolean, and all values are NaNs.
```
<Column name='is_star' dtype='float64' description='Flag indicating whether the source is a star' length=116>
nan
nan
nan
nan

```
Do you know if this column is computed in photutils or source_catalog?

nden replied:
From `source_catalog`:
```
@lazyproperty
    def is_star(self):
        """
        Flag indicating whether the source is a star.

        2020.03.02: The JWST Photometry Working Group has not determined
        the criteria for this flag.
        """

        # TODO: need algorithm for this flag
        #is_star = np.random.randint(2, size=len(self.id))
        #return is_star.astype(bool)
        return self.null_column
```

This is where NaNs are returned.

hbushouse replied:
An algorithm was specified a month or two ago, but it hasn't been implemented yet. It will be Boolean when implemented, with a value of `True` for point sources.

hbushouse commented on jwst/assign_wcs/util.py:
```diff
@@ -521,19 +523,32 @@ def create_grism_bbox(input_model,
+        Cross-dispersion extraction half height in pixels, WFSS mode.
+        Overwrites the computed extraction height in ``GrismObject.order_bounding.``
+        If ``None``, it's computed from the segementation map,
+        using the min and max wavelegnth for each of the orders that
```
typo in "wavelegnth"

hbushouse commented on jwst/assign_wcs/util.py:
```diff
@@ -567,49 +580,55 @@ def create_grism_bbox(input_model,
+            message = "If reference files are not supplied, ``wavelength_range`` must be provided."
+            raise TypeError(message)
+    else:
+        # Get the list of extract_orders and lmin, lmax from the ``wavelengthrange11 reference file.
```
typo "wavelengthrange11"

hbushouse commented on jwst/assign_wcs/util.py:
```diff
@@ -567,49 +580,55 @@ def create_grism_bbox(input_model,
+    return _create_grism_bbox(input_model, mmag_extract, wfss_extract_half_height, wavelength_range)
+
+
+def _create_grism_bbox(input_model, mmag_extract=99.0,
```
Do we want mmag hardwired to 99 or allow a user value to be passed in?

nden replied:
The default value is 99, a user can change it.

hbushouse commented on jwst/extract_2d/extract_2d_step.py:
```diff
@@ -16,7 +16,8 @@ class Extract2dStep(Step):
         slit_name = string(default=None)
         extract_orders = int_list(default=None)  # list of orders to extract
-        extract_height =  integer(default=None)  # extraction height in pixels
+        extract_height =  integer(default=None)  # extraction height in pixels, TSGRISM mode
```
Since we now have 2 extraction height params, maybe we should be explicit in the names for both of them which mode they apply to, such as `tsgrism_extract_height`.

hbushouse commented on jwst/extract_2d/grisms.py:
```diff
@@ -299,6 +300,9 @@ def extract_grism_objects(input_model,
         Compute a wavelength array for the datamodel.  Computationally
         expensive, but saves doing it repeatedly later on in pipeline.
 
+    wfss_extract_half_heigh : int, (optional)
```
typo: missing the trailing "t" on height

hbushouse commented on jwst/assign_wcs/util.py:
```diff
@@ -658,6 +677,17 @@ def create_grism_bbox(input_model,
                     xmax = int(np.max(xstack))
                     ymin = int(np.min(ystack))
                     ymax = int(np.max(ystack))
+                    if wfss_extract_half_height is not None:
```
Is the source catalog info for each source known at this point in the processing? This param value needs to be applied ONLY if the source being processed has `is_star = True` in the source catalog (only applied to point sources).

nden replied:
yes, the source catalog is available, will include this check

nden commented:
I added the check for `is_star` with a made up value. This will need to be revised after the decision of how to populate the column. I also addressed the rest of the comments.

hbushouse commented on jwst/assign_wcs/util.py:
```diff
@@ -659,6 +680,21 @@ def create_grism_bbox(input_model,
+                    # The ``72`` criteria is a placeholder fof when there's a decision
+                    # on how to populate the ``is_star`` column in the catalog.
+                    # See ``is_star`` property in source_catalog.py
+                    if wfss_extract_half_height is not None and obj.is_star > 72:
```
Is there a way to test for `NaN` versus `True`, such that if `NaN`, don't apply, and if `True`, then do apply it.

hbushouse commented on docs/jwst/extract_2d/main.rst:
```diff
@@ -133,11 +208,13 @@ For NIRCam and NIRISS WFSS, the ``extract_2d`` step has three optional arguments
-  reference file.
+For NIRCam TSGRISM, the ``extract_2d`` step has one optional argument:
 
 ``--extract_height``
```
Should update this now to ``--tsgrism_extract_height``

hbushouse commented on jwst/assign_wcs/tests/data/step_SourceCatalogStep_cat.ecsv:
```diff
@@ -23,6 +23,7 @@
 # - {name: orientation_sky, unit: deg, datatype: float64}
 # - {name: isophotal_abmag, datatype: float64}
 # - {name: isophotal_abmag_err, datatype: float32}
+# - {name: is_star, datatype: float64}
```
Perhaps you should modify the data file used in the test to have the `is_star` column be Boolean and modify the entries in the ecsv file itself to also be boolean for that column (perhaps with a mixture of True and False values).

hbushouse commented on jwst/assign_wcs/util.py:
```diff
@@ -659,6 +680,21 @@ def create_grism_bbox(input_model,
                     ymin = int(np.min(ystack))
                     ymax = int(np.max(ystack))
 
+                    # The ``72`` criteria is a placeholder fof when there's a decision
```
You can get rid of this comment now, since you've already changed `obj.is_star` to be used as a boolean.

hbushouse commented on jwst/lib/catalog_utils.py:
```diff
@@ -118,6 +119,8 @@ class SkyObject(namedtuple('SkyObject', ("id",
         Upper left corner of the minimum bounding box
     sky_bbox_ur : `~astropy.coordinates.SkyCoord`
         Upper right corner of the minimum bounding box
+    is_star : float
```
Need to change the data type to `bool`
